Bug Fixes and add param to savedonor

RonAbarbanel opened the pull request:
1) Fix: Don't allow + symbol in params as it generates an error.
2) Fix: + symbol resolves to a numeric but is not numeric (ex. in phone numbers). Treat anything with a + as a string so the + can be removed.
3) Fix: Card expiration date resolves to a numeric field (in vault) but the data field is not numeric.
4) Fix: nomail is a required field and generates an error if set to NULL. If unspecified, set it to N.
5) Add: receipt_delivery param to savedonor (not in the API documentation, but a legitimate parameter)

RonAbarbanel committed:
Bug Fixes and add param to savedonor

1) Fix: Don't allow + symbol in params as it generates an error.
2) Fix: + symbol resolves to a numeric but is not numeric (ex. in phone numbers). Treat anything with a + as a string so the + can be removed.
3) Fix: Card expiration date resolves to a numeric field (in vault) but the data field is not numeric.
4) Fix: nomail is a required field and generates an error if set to NULL. If unspecified, set it to N.
5) Add: receipt_delivery param to savedonor

LukeTowers commented on src/DonorPerfect.php:
```diff
@@ -628,6 +629,11 @@ public function dp_donorsearch($data)
      */
     public function dp_savedonor($data)
     {
+        // The nomail parameter is a required field for dp_savedonor. Make sure it is
+        // set properly and default to N if missing or invalid to prevent error.
+        if (!isset($data['nomail']) || $data['nomail'] != 'Y') $data['nomail'] = 'N';
+        // receipt delivery defaults to L if unspecified, make it explicit
+        if (!isset($data['receipt_delivery'])) $data['receipt_delivery'] = 'L';
```
```suggestion
        // nomail is required for dp_savedonor, ensure it is present and valid
        if (!isset($data['nomail']) || $data['nomail'] != 'Y') {
            $data['nomail'] = 'N';
        }
        // receipt_delivery defaults to L if unspecified, make it explicit
        if (!isset($data['receipt_delivery'])) {
            $data['receipt_delivery'] = 'L';
        }

```

LukeTowers commented on src/DonorPerfect.php:
```diff
@@ -218,8 +218,9 @@ public function call(string $action, array $parameters)
+                // + in param will be interpreted as a space and will generate an error
                 // @see https://api.warrenbti.com/2020/08/03/apostrophes-in-peoples-names/
-                $value = str_replace(["'", '"', '%'], ["''", '', '%25'], $value);
+                $value = str_replace(["'", '"', '%', '+'], ["''", '', '%25', ''], $value);
```
```suggestion
                $value = str_replace(["'", '"', '%', '+'], ["''", '', '%25', '%2b'], $value);
```
We should probably just URL encode the `+` rather than remove it outright

LukeTowers replied:
@RonAbarbanel any thoughts?

LukeTowers commented on src/DonorPerfect.php:
```diff
@@ -208,7 +207,8 @@ public function call(string $action, array $parameters)
             $value = trim($value);
 
-            if (is_numeric($value) && strpos($value, 'e') === false) {
+            if (is_numeric($value) && strpos($value, 'e') === false && strpos($value, '+') === false &&
+               $param != 'CardExpirationDate') {
```
```suggestion
            if (is_numeric($value)
                && strpos($value, 'e') === false
                && strpos($value, '+') === false
                && $param != 'CardExpirationDate'
            ) {
```

LukeTowers replied:
@RonAbarbanel could you provide me the value of CardExpirationDate that is causing issues being interpreted as a numeric and then what it should actually be processed as?

I'd like to take a stab at fixing that through recognition of the value, rather than the param name since there is likely other params with similar behaviour.

LukeTowers committed:
Update src/DonorPerfect.php

LukeTowers commented:
Thanks for the contribution @RonAbarbanel! Provided some feedback, let me know your thoughts :)

RonAbarbanel commented:
Hi Luke,

The CardExpirationDate is stored in the DP Accounts (Vault) and is in the
form MMYY. If it is a numeric, the API will throw an error, it must be
treated as a string. So for instance 0225 should be '0225' not 225.

You are correct that there may be other parameters that behave the same
way. So far I have not run into any. But your method would be superior.

Ron

On Wed, Jan 4, 2023 at 1:22 PM Luke Towers ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/DonorPerfect.php
> <https://github.com/LukeTowers/php-donorperfect-api/pull/2#discussion_r1061715770>
> :
>
> > +            if (is_numeric($value) && strpos($value, 'e') === false && strpos($value, '+') === false &&
> +               $param != 'CardExpirationDate') {
>
> @RonAbarbanel <https://github.com/RonAbarbanel> could you provide me the
> value of CardExpirationDate that is causing issues being interpreted as a
> numeric and then what it should actually be processed as?
>
> I'd like to take a stab at fixing that through recognition of the value,
> rather than the param name since there is likely other params with similar
> behaviour.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/LukeTowers/php-donorperfect-api/pull/2#discussion_r1061715770>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEXDFYFSG6X47DXVCSUZ7KTWQWWUFANCNFSM6AAAAAATKXJENY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

RonAbarbanel commented:
Ah, I see the problem. You are encoding things twice - so changing the + to
%2B results in %252B in the final call.

        // Validate the API call before making it
        $relativeUrl .= http_build_query($args, null, '&',
PHP_QUERY_RFC3986); (this is where you double encode)

I will add this instead immediately following the above to deal with the +
sign.
        // encode any + signs
        $relativeUrl = str_replace('+', '%2B', $relativeUrl);

This also turned up another small possibly insignificant problem where you
use str_ireplace - this will still leave many duplicate spaces (like when
there are 4 spaces in a row). I decided to fix it though and will add an
update.

  // Clean the sql of extra spaces and tabs
        $cleansql = $sql;
        $done = false;
        do {
          $cleansql_new = trim(str_ireplace(["\n", "\t", '  ', '  )'], ['
', '', ' ', ' )'], $cleansql));
          if ($cleansql_new === $cleansql) {
            $done = true;
          }
          else {
            $cleansql = $cleansql_new;
          }
        } while (!$done);

        $params = [
            'action' => $cleansql_new,
        ];

Make sense?
Ron

On Wed, Jan 4, 2023 at 1:22 PM Luke Towers ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/DonorPerfect.php
> <https://github.com/LukeTowers/php-donorperfect-api/pull/2#discussion_r1061716207>
> :
>
> >                  // @see https://api.warrenbti.com/2020/08/03/apostrophes-in-peoples-names/
> -                $value = str_replace(["'", '"', '%'], ["''", '', '%25'], $value);
> +                $value = str_replace(["'", '"', '%', '+'], ["''", '', '%25', ''], $value);
>
> @RonAbarbanel <https://github.com/RonAbarbanel> any thoughts?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/LukeTowers/php-donorperfect-api/pull/2#discussion_r1061716207>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEXDFYALGMWNUURPFA6FTNDWQWWVDANCNFSM6AAAAAATKXJENY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

RonAbarbanel committed:
Code cleanup and ...

Add currency and acknowledge pref to savepledge

RonAbarbanel committed:
Update DonorPerfect.php

RonAbarbanel committed:
Update DonorPerfect.php

LukeTowers commented on src/DonorPerfect.php:
```diff
@@ -250,8 +255,21 @@ public function call(string $action, array $parameters)
      */
     public function callSql($sql)
     {
+        // Clean the sql of extra spaces and tabs
+        $cleansql = $sql;
+        $done = false;
+        do {
+          $cleansql_new = trim(str_ireplace(["\n", "\t", '  ', '  )'], [' ', '', ' ', ' )'], $cleansql));
+          if ($cleansql_new === $cleansql) {
+            $done = true;
+          }
+          else {
+            $cleansql = $cleansql_new;
+          }
+        } while (!$done);
+
```
```suggestion
        // Remove all formatting whitespace while leaving whitespace that is part of value strings
        $in_quote = false;
        $output = '';

        for ($i = 0; $i < strlen($sql); $i++) {
            if ($sql[$i] == "'" || $sql[$i] == '"') {
                $in_quote = !$in_quote;
            }
            if (($sql[$i] == ' ' || $sql[$i] == "\n" || $sql[$i] == "\r") && !$in_quote) {
                if (empty($output) || substr($output, -1) == ' ') {
                    continue;
                }
                $output .= ' ';
            } else {
                $output .= $sql[$i];
            }
        }
        
```

LukeTowers replied:
@RonAbarbanel what do you think of the above? Kindly provided by ChatGPT after about 15 minutes of back and forth, it takes

```php
$sql = "SELECT *
       FROM my_table
       WHERE col1 = 'val1
                   val1_cont
                   val1_cont'
         AND col2 = \"val2
                   val2_cont
                   val2_cont\"";
```

and turns it into 

```sql
SELECT * FROM my_table WHERE col1 = 'val1
                   val1_cont
                   val1_cont' AND col2 = "val2
                   val2_cont
                   val2_cont"
```

whereas the same input using the do while loop resulted in

```sql
SELECT * FROM my_table WHERE col1 = 'val1 val1_cont val1_cont' AND col2 = "val2 val2_cont val2_cont"
```

LukeTowers replied:
@jaxwilko @bennothommo FYI my first use of ChatGPT on a programming problem lol 😂

RonAbarbanel replied:
Haha, this is fantastic. I did not consider the in-quote possibility. I guess ChatGPT is smarter than me. I'll have to start using it!

RonAbarbanel committed:
Update src/DonorPerfect.php

Co-authored-by: Luke Towers <github@luketowers.ca>

RonAbarbanel committed:
set sql action to correct variable after new clean procedure

LukeTowers commented on src/DonorPerfect.php:
```diff
@@ -758,6 +791,8 @@ public function dp_savepledge($data)
             'receipt_delivery_g'  => ['string', 1], // ‘E’ for email, ‘B’ for both email and letter, ‘L’ for letter, ‘N’ for do not acknowledge or NULL
             'contact_id'          => ['numeric'], // Or NULL
+            'acknowledgepref'     => ['string', 3],
+            'currency'            => ['string', 3]
```
```suggestion
            'currency'            => ['string', 3],
```

LukeTowers committed:
Update src/DonorPerfect.php

LukeTowers committed:
Formatting & improvements to the trimSql() method

See below for the list of statements this was tested against:

```php
// Test cases
$tests = [
    // No quotes
    'SELECT * FROM users' => [
        // Unchanged
        'SELECT * FROM users',
        // Trim start, end, and in between
        '   SELECT    *    FROM   users   ',
        // Trim start, end, and in between, and remove newlines
        "\t\tSELECT \n\t\t\t*\n\t\tFROM\n\t\t\tusers\n\t\t",
        // Trim SQL formatted for readability
        '
            SELECT 
                *
            FROM
                users
        ',
    ],
    // Double quotes
    'SELECT * FROM USERS WHERE name = "BOB  BUILDER"' => [
        // Unchanged
        'SELECT * FROM USERS WHERE name = "BOB  BUILDER"',
        // Trim start, end, and in between
        '   SELECT    *    FROM   USERS   WHERE   name   =   "BOB  BUILDER"   ',
        // Don't remove whitespace inside double quotes
        "SELECT * FROM USERS WHERE name = \"BOB  BUILDER\"",
    ],
    // Single quotes
    'SELECT * FROM USERS WHERE name = \'BOB  BUILDER\'' => [
        // Unchanged
        'SELECT * FROM USERS WHERE name = \'BOB  BUILDER\'',
        // Trim start, end, and in between
        '   SELECT    *    FROM   USERS   WHERE   name   =   \'BOB  BUILDER\'   ',
        // Don't remove whitespace inside single quotes
        "SELECT * FROM USERS WHERE name = 'BOB  BUILDER'",
    ],
    // Escaped quotes
    "SELECT * FROM USERS WHERE name = \"BOB \\\" BUILDER" => [
        // Don't remove whitespace inside double quotes, even if escaped
        "SELECT * FROM USERS WHERE name = \"BOB \\\" BUILDER",
    ],
];
```

LukeTowers committed:
Wrap string values in an anonymous class

LukeTowers commented:
@RonAbarbanel what do you think of [`3920a8c` (#2)](https://github.com/LukeTowers/php-donorperfect-api/pull/2/commits/3920a8c1d4d9b1a803f7cdf24f2cab7eeaf92562) as a solution to the Card Expiration Date being parsed as a numeric instead of a string?

LukeTowers commented on src/DonorPerfect.php:
```diff
@@ -121,10 +121,12 @@ protected function callInternal(array $params)
         // Validate the API call before making it
         $relativeUrl .= http_build_query($args, null, '&', PHP_QUERY_RFC3986);
+        // encode any + signs
+        $relativeUrl = str_replace('+', '%2B', $relativeUrl);
```
Could you provide the exact input / call that you were making that was causing this issue with the plus sign not being properly encoded?

RonAbarbanel replied:
$dpAPI->dp_savedonor(array( ... , 'mobile_phone' => '+19028307662'));

LukeTowers replied:
Hmm, I'll need you to test this some more, I ran a simplified version of this to test and it seems to be encoding the `+` correctly, see below:

```php
/**
 - dp_savedonor
    - prepareParams
      - prepareString
    - call
 */

function call(string $action, array $parameters)
{
    $paramString = '';
    $i = 0;
    foreach ($parameters as $param => $value) {
        $original = $value;
        $value = trim($value);

        if (
            !is_object($original)
            && is_numeric($value)
            && strpos($value, 'e') === false
            && strpos($value, '+') === false
        ) {
            $value = $value;
        } elseif (is_bool($value)) {
            $value = $value ? '1' : '0';
        } elseif (empty($value)) {
            $value = 'NULL';
        } elseif (is_array($value)) {
            $value = "N'" . implode($value, '|') . "'";
        } else {
            // Ensure quotes are doubled for escaping purposes
            // @see https://api.warrenbti.com/2020/08/03/apostrophes-in-peoples-names/
            $value = str_replace(["'", '"', '%'], ["''", '', '%25'], $value);

            // Wrap the value in quotes
            $value = "'$value'";
        }

        $paramString .= '@'.$param.'='.$value;
        ++$i;
        if ($i !== count($parameters)) {
            $paramString .= ',';
        }
    }

    $params = [
        'action' => $action,
        'params' => $paramString,
    ];

    return $params;
}


$params = [
    'mobile_phone' => new class('+19028307662') {
            protected string $value;
            public function __construct(string $value)
            {
                $this->value = $value;
            }
            public function __toString()
            {
                return $this->value;
            }
        },
];

$prepared = call('dp_savedonor', $params);

$result = http_build_query($prepared, null, '&', PHP_QUERY_RFC3986);
```

Which generates `action=dp_savedonor&params=%40mobile_phone%3D%27%2B19028307662%27`

RonAbarbanel commented:
> @RonAbarbanel what do you think of [`3920a8c` (#2)](https://github.com/LukeTowers/php-donorperfect-api/pull/2/commits/3920a8c1d4d9b1a803f7cdf24f2cab7eeaf92562) as a solution to the Card Expiration Date being parsed as a numeric instead of a string?

I think this is a better solution as it forces the API to adhere to the type set out in the parameter. I have not tested it, but eyeballing the code looks good. I will run some tests tomorrow.

LukeTowers committed:
Ensure that class wrapped values aren't interpreted as strings

LukeTowers committed:
Only trim strings

LukeTowers commented:
@RonAbarbanel any luck with the tests?

RonAbarbanel commented:
> @RonAbarbanel any luck with the tests?

@LukeTowers Yes ... I just finished and everything is purring like a kitten. I removed the extra check for the '+' as it is not needed. I think that the previous issue was that it was being treated like a number and causing an error. Now that we check the phone field and treat it like text, there is no issue. I will push that change. I don't foresee any other changes for the time being, So this probably will conclude my changes in the near future.

RonAbarbanel committed:
Remove extra encoding of the + (not needed)

LukeTowers committed:
Update DonorPerfect.php
